### ISSUE TITLE
Add support for xforms-value-changed event

### ISSUE THREAD

**Issue by lognaturel** [opendatakit/javarosa#227]
There's been agreement to add actions and events more formally to the ODK XForms spec -- https://github.com/opendatakit/xforms-spec/issues/123#issuecomment-348580247. In particular, the `xforms-value-changed` event is now part of the spec. This should be supported such that actions with the `event` attribute set to `xforms-value-changed` can be nested in form control elements in the body.

Sample form: [lightweight-timestamping.xml.txt](https://github.com/opendatakit/javarosa/files/1523249/lightweight-timestamping.xml.txt)

First thing to figure out is exactly when the event will dispatched. Similarly, I think there will be some decisions to be made around when dependent triggerables are reevaluated when the `setvalue` action is used. Or maybe the two are the same question, I'm not sure yet.

(

<strike>See section 4.4.2 at https://www.w3.org/TR/2006/REC-xforms-20060314/slice4.html#evt-notify:
> For incremental processing, this specification does not define how often XForms Processors fire these events. Implementations are expected to optimize processing (for instance not flashing the entire screen for each character entered, etc.).</strike> 

The above is only relevant if the `incremental` attribute is true and we don't support that currently. From https://www.w3.org/community/xformsusers/wiki/XForms_2.0#Common_attribute_and_content_sets:
> Normally, controls that only require a single interaction update the instance immediately after a change; controls that require more interaction, like text-oriented controls such as textarea, only update the instance once focus is moved away. When false, this attribute indicates that the instance data should be updated when the control loses focus; when true, the instance data should be updated more often, thereby generating more xforms-value-changed events – however, an implementation may buffer sequences of changes happening in quick succession, and apply them all at once. The default value for this attribute varies per control and modality, but for textual controls the default is normally false.

)

**Comment by lognaturel** [comment]
Currently `FormDef` has an `eventListeners` map mapping events to `Action`s. Should `xforms-value-changed` events be kept track of there? Or maybe on form controls themselves?

**Comment by mdudzinski** [comment]
@lognaturel There is no such concept as focus in JavaRosa so I think this event should be dispatched at the same level as DAG stuff is fired and focus stuff should be left to a client app. I've been thinking about `FormDef.setValue(IAnswerData data, TreeReference ref, TreeElement node, boolean midSurvey)` method. This would be consistent with other implemented events. For instance `initialize(boolean newInstance, InstanceInitializationFactory factory)` method which dispatches `EVENT_XFORMS_READY` and also calls `initializeTriggerables`. I realize that those events are not actually used by Collect and most likely were merged along with some Dimagi patches but since those it for them then then it may work for us.

**Comment by lognaturel** [comment]
Setvalue sounds right. This is a value-changed event so even if it were possible to use focus that wouldn’t be right.

Probably the event should only fire if the value has actually changed, not if the same on is saved again?

I think event-XForms-ready works without any additional client support. That would be like the form with setvalue that you added as a test recently, right?

**Comment by mdudzinski** [comment]
@lognaturel Exactly. Ideally we should check if the value has actually changed to avoid unnecessary evaluation of triggerables. But I think it can be added later. As we discussed it some time ago - it won't be hard to do but will require some time to do it right (clean code + tests). 

The biggest problem is the support for the nested `setvalue` action but I've already got some idea. Will require a bit of refactoring in the existing action handling (but there are comments saying about future changes so obviously the current code was just a first step anyway) and changes in the inputs parsing. We need to register an action there.
